File: Binary Search/Step 2/F. String Game.cpp
#include <bits/stdc++.h>
using namespace std;

string t, p;
vector<int> orders;

bool isIn(string s1, string s2) {
    int i = 0;
    for(char c: s1) {
        if (i < s2.size() && s2[i] == c) i++;
    }
    return i == (int)s2.size();
}

bool f(int n) {
    string curr = "";
    
    unordered_set<int> removable;
    for(int i =0; i < n; i++) {
        removable.insert(orders[i]);
    }

    for(int i = 0; i < t.size(); i++) {
        if(removable.find((i+1)) == removable.end()) {
            curr += t[i];
        }
    }

    return isIn(curr, p);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    cin >> t >> p;
    orders.resize(t.size());
    for(int i=0; i < t.size(); i++) {
        cin >> orders[i];
    }

    long long l = 0;
    long long r = t.size();
    while(l+1<r) {
        auto m = (l+r)/2;
        if(f(m)) l = m;
        else r = m;
    }

    cout << l << "\n";

    return 0;
}

File: Binary Search/Step 2/F. String Game.py
t = input()
p = input()
orders = list(map(int, input().split()))

def s2Ins1(s1, s2):
    if len(s2) > len(s1):
        return False
    i = 0
    for char in s1:
        if i < len(s2) and char == s2[i]:
            i += 1
    return i == len(s2)

def f(n):
    curr = ""
    removable = set()
    for i in range(n):
        removable.add(orders[i])
    for i in range(len(t)):
        if (i+1) not in removable:
            curr += t[i]
    return s2Ins1(curr, p)
        


l = -1
r = len(orders)

while l+1 < r:
    m = (l+r)//2
    if f(m):
        l = m
    else:
        r = m

print(l)